C++ example documentation

kalaspuffar opened the pull request:
I was trying to figure out the examples for Bazel and found that the documentation for the first steps was missing so I figured I start the conversation by writing the first copy.

bazel-io commented:
Can one of the admins verify this patch?

damienmg commented:
Unfortunately I just realized that nobody is watching PR for that repository. sorry for the delay. NYC is offline for the weekend but they should come back at you in a few days.

Sorry again.

jin commented on cpp-tutorial/README.md:
```diff
@@ -0,0 +1,12 @@
+# Examples to build c++ code
+
+This package will showcase how to build c++ code in stages.
```
nit: C++

jin commented on cpp-tutorial/stage1/README.md:
```diff
@@ -0,0 +1,31 @@
+
+This showcases how to build a single file to create a runnable application.
+
+In the build file for this stage you can read that this will use the c-compiler to build a binary file. (cc_binary)
```
nit: BUILD file

We can't really infer this from the BUILD file.. here we can say instead "This BUILD file shows that we want to build a C++ binary using the `cc_binary` rule provided by Bazel.`

jin commented on cpp-tutorial/stage1/README.md:
```diff
@@ -0,0 +1,31 @@
+This showcases how to build a single file to create a runnable application.
+
+In the build file for this stage you can read that this will use the c-compiler to build a binary file. (cc_binary)
+The name of the binary will be hello-world
```
backticks around `hello-world`

jin commented on cpp-tutorial/stage1/README.md:
```diff
@@ -0,0 +1,31 @@
+
+In the build file for this stage you can read that this will use the c-compiler to build a binary file. (cc_binary)
+The name of the binary will be hello-world
+And we also name the required source files to build the current application.
```
"In the `cc_binary` rule, let's name the binary `hello-world`, and provide the required source files in the `srcs` attribute."

jin commented on AUTHORS:
```diff
@@ -8,3 +8,4 @@
 
 Google Inc.
 Manfred Touron <m@42.am>
+Daniel Persson <mailto.woden@gmail.com>
```
please rm

jin commented on cpp-tutorial/README.md:
```diff
@@ -0,0 +1,12 @@
+This package will showcase how to build c++ code in stages.
+
+### Stage 1
+The first stage is really simple and shows you how to build a single file.
```
rm "is really simple"

"how to build a single file" -> "how to compile a binary with a single source file"

jin commented on cpp-tutorial/README.md:
```diff
@@ -0,0 +1,12 @@
+The first stage is really simple and shows you how to build a single file.
+
+### Stage 2
+The second stage will showcase how to build an application with multiple files, separated in a library and a binary.
```
s/multiple files/multiple source and header files/

jin commented on cpp-tutorial/README.md:
```diff
@@ -0,0 +1,12 @@
+The second stage will showcase how to build an application with multiple files, separated in a library and a binary.
+
+### Stage 3
+The third stage showcases how to link multiple build directories building multiple libraries in different packages and then connecting it up with the main application.
```
s/building/by building/

jin commented on cpp-tutorial/stage1/README.md:
````diff
@@ -0,0 +1,31 @@
+bazel build //main:hello-world
+```
+
+The build will output some information for example
````
If the build is successful, Bazel prints the following output:

jin commented on cpp-tutorial/stage1/README.md:
````diff
@@ -0,0 +1,31 @@
+
+To build this example you use (notice that 3 slashes are required in windows)
+```
+bazel build //main:hello-world
````
Let's also show the Windows example:

```python
bazel build //main:hello-world

# In Windows, note the three slashes

bazel build ///main:hello-world
```

jin commented on cpp-tutorial/stage2/README.md:
```diff
@@ -0,0 +1,30 @@
+
+### Library
+
+Below we see a new configuration option. This is a c-compiler library with a seperate header file as an interface, the sources and a name for the library to be built.
```
`cc_library` is not a configuration option, it's a rule. 

You can say: "Here, we introduce the `cc_library` rule for building C++ libraries. Here, we have a `cc_library` named `hello-greet` and its header and source files are defined accordingly."

jin replied:
Sorry, small fix: can you please remove the "Here," from the second sentence?

jin commented on cpp-tutorial/stage2/README.md:
```diff
@@ -0,0 +1,30 @@
+
+### Binary
+
+The binary configuration we saw in stage 1 has not changed except that we now depend on the library hello-great.
```
nit: Stage 1

nit: s/binary/`cc_binary`/

nit: ..has not changed, except that..

nit: s/library/`cc_library`

nit: s/hello-great/`hello-greet`

jin commented on cpp-tutorial/stage2/README.md:
````diff
@@ -0,0 +1,30 @@
+)
+```
+
+To build this example you use (notice that 3 slashes are required in windows)
````
can rm the Windows slashes line, since it's mentioned in stage 1

jin commented on cpp-tutorial/stage3/README.md:
```diff
@@ -0,0 +1,30 @@
+# Stage 3
+
+In this stage we step it up and showcase how to integrate multiple libraries and even in different packages.
```
"how to integrate multiple `cc_library` targets from different packages"

jin commented on cpp-tutorial/stage3/README.md:
```diff
@@ -0,0 +1,30 @@
+
+In this stage we step it up and showcase how to integrate multiple libraries and even in different packages.
+
+Below we see a similar configurtion from stage 2. The differance is that this BUILD file is in the lib package / directory. And the new property below visibility will tell bazel which other packages that should be able to see / use this package. So in this case the main package can use this library. 
```
s/configurtion/configuration/

s/differance/difference/

The phrasing sounds a little off here, how about this? "Below, we see a similar configuration from Stage 2, except that this BUILD file is in a subdirectory called `lib`. In Bazel, subdirectories containing BUILD files are known as packages."

The official term for something like `visibility` is an attribute, not property. 

rm "And" from "And the new.."

s/bazel/Bazel

s/main package/`//main` package

replace "see / use" with "reference"

jin commented on cpp-tutorial/stage3/README.md:
````diff
@@ -0,0 +1,30 @@
+)
+```
+
+Then to use the library above we add an extra dependency for //[package-path]:[dependency-name].
````
s/for/in the form of

s/[dependency-name]/[target-name]

jin commented on cpp-tutorial/stage3/README.md:
````diff
@@ -0,0 +1,30 @@
+)
+```
+
+To build this example you use (notice that 3 slashes are required in windows)
````
rm note about the three slashes, since it's mentioned in Stage 1

kalaspuffar commented:
Hi @jin.

Thank you for the suggested changes. I've looked through them and implemented all as changes to this PR.

I'm not sure if the suggestion to remove the information about the differences in build steps for windows and Linux is a bad thing to have in all the tutorial steps. Sometimes people read them out of order and miss some crucial information. Then again I think it will be apparent when running the build.

Best regards
Daniel

jin commented:
SGTM on leaving the Linux/Windows note in.

I've also sent in my comments on Stage 3, please take a look.

kalaspuffar commented:
Hi @jin.

I've added the full build instructions back to both Stage 2 and 3 and removed Here, in the second sentence.

Thanks for your time, and please tell me if I've missed or misunderstood something.

Best regards
Daniel

jin commented:
Thanks for the updates! I think you missed some here: https://github.com/bazelbuild/examples/pull/41/files/66ae6016c793e6f3b2f90f457eb28644b3e4fdec#diff-f3733191e70b5b78a58a0b17a2020f94

jin commented:
LGTM, @spomorski do you have any inputs?

kalaspuffar commented:
Hi @jin 

I've missed some backticks at one place and also changed so I use the wording Rule everywhere instead of configuration.

Thank you for your time.

Best regards
Daniel
